machineconfiguration/v1/types_machineosbuild.go

```diff
@@ -0,0 +1,217 @@
+type MachineOSBuilderReference struct {
+	// imageBuilderType describes the type of image builder used to build this image.
+	// Valid values are Job only.
+	// When set to Job, a pod based builder, using buildah, is launched to build the specified image.
+	// +unionDiscriminator
+	// +required
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+
+	// job is a reference to the job object that is managing the image build.
+	// This is required if the imageBuilderType is Job, and forbidden otherwise.
+	// +unionMember
+	// +optional
+	Job *ObjectReference `json:"job,omitempty"`
+}
+
+// BuildProgess highlights some of the key phases of a build to be tracked in Conditions.
+type BuildProgress string
+
+const (
+	// prepared indicates that the build has finished preparing. A build is prepared
+	// by gathering the build inputs, validating them, and making sure we can do an update as specified.
+	MachineOSBuildPrepared BuildProgress = "Prepared"
+	// building indicates that the build has been kicked off with the specified image builder
+	MachineOSBuilding BuildProgress = "Building"
+	// failed indicates that during the build or preparation process, the build failed.
+	MachineOSBuildFailed BuildProgress = "Failed"
+	// interrupted indicates that the user stopped the build process by modifying part of the build config
+	MachineOSBuildInterrupted BuildProgress = "Interrupted"
+	// succeeded indicates that the build has completed and the image is ready to roll out.
+	MachineOSBuildSucceeded BuildProgress = "Succeeded"
+)
+
+// Refers to the name of a rendered MachineConfig (e.g., "rendered-worker-ec40d2965ff81bce7cd7a7e82a680739", etc.):
+// the build targets this MachineConfig, this is often used to tell us whether we need an update.
+type MachineConfigReference struct {
+	// name is the name of the rendered MachineConfig object.
+	// This value should be between 10 and 253 characters, and must contain only lowercase
+	// alphanumeric characters, hyphens and periods, and should start and end with an alphanumeric character.
+	// +kubebuilder:validation:MinLength:=10
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +required
+	Name string `json:"name"`
+}
+
+// ObjectReference contains enough information to let you inspect or modify the referred object.
+type ObjectReference struct {
+	// group of the referent.
+	// The name must contain only lowercase alphanumeric characters, '-' or '.' and start/end with an alphanumeric character.
+	// Example: "", "apps", "build.openshift.io", etc.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
```

Comment:
Rather than `Pattern`, use `XValidation` with `self.matches(...)` and add the same error message as in 4.19

Reply:
Fixed